Added PressurePlate set, HitSwitch and SwitchTiles hooks

Tyfyter opened the pull request:
### What is the new feature?
Added `TileID.Sets.PressurePlate`.
Added `(Mod|Global)Tile.HitSwitch` and `(Mod|Global)Tile.SwitchTiles`.

### Why should this be part of tModLoader?
Currently every mod that adds a pressure plate has to use IL editing to do so, multiple mods which add pressure plates are extremely likely to conflict.

Additionally, there are currently no hooks in `Collision.SwitchTiles` allowing the creation of tiles that react to entity collisions.

Finally, a hook in `Wiring.HitSwitch` is also missing. While only specific content calls this method, it is extremely convenient for implementing tile interaction behaviors that should run on all clients and the server. By adding a hook here, tiles can sync effects without writing custom `ModPacket`s.

### Sample usage for the new feature
Custom pressure plate tile:
```cs
public class MyPressurePlateTile : ModTile {
	public override void SetStaticDefaults() {
		…
		TileID.Sets.PressurePlate[Type] = 0; // Reacts to all entities. Different values do different things, see the documentation.
		…
	}
}
```

Using `HitSwitch` instead of `HitWire` or `RightClick` for effects that should run on all clients and the server.
```cs
public class MyInteractionTile : ModTile {
	public override bool RightClick(int i, int j) {
		Wiring.HitSwitch(i, j);
		NetMessage.SendData(MessageID.HitSwitch, number: i, number2: j);
	}

	public override void HitWire(int i, int j) {
		Wiring.HitSwitch(i, j);
		NetMessage.SendData(MessageID.HitSwitch, number: i, number2: j);
	}

	public override void HitSwitch(int i, int j) {
		Wiring.TripWire(i, j, 1, 1); 
		Dust.NewDust(new Vector2(i * 16, j * 16), 16, 16, DustID.GreenFairy);
		SoundEngine.PlaySound(SoundID.Unlock, new Vector2(i * 16, j * 16));
	}
}
```

### ExampleMod Updates
`ExamplePressurePlate` has been added. It demonstrates a weighted pressure plate but also has comments showing how to convert to a normal pressure plate tile.

`ExampleSlopeTile` has been updated to be a one-way sensor, sending a wire signal when a player passes through it in the indicated direction. This example shows off the new `ModTile.SwitchTiles` hook and demonstrates custom entity collision logic with a tile.

Here is a video of the new behavior of `ExampleSlopeTile`:

https://github.com/user-attachments/assets/5c1a9d64-381a-49c1-a668-037bc6cd94eb

### Discussion
- `ExampleSlopeTile` can only place on solid tiles, making it a bit difficult to place if intended for puzzle gates. It might be worth moving the example to a more appropriately named tile and use `TileID.Sets.CanPlaceNextToNonSolidTile` like `ExampleLivingFireTile` does.
- The new hook names match the vanilla method names where they are located, which is helpful for modders seeking to adapt vanilla code. `SwitchTiles`, however, isn't really indicative of what it does, and is plural which is not what we usually do.
- `TileID.Sets.PressurePlate` value options might seem a bit strange, but I wanted to easily support all vanilla options without requiring modders to write collision checking logic for a simple pressure plate. It might be better as a `int?[]` rather than a `int[]`.
- ~~Should we add a `Wiring.HitSwitchAndSync` helper method? `HitSwitch` is almost always followed by `NetMessage.SendData(MessageID.HitSwitch...)`.~~ - Added

### Porting Notes
- Replace IL edits with setting the appropriate pressure plate type in `TileID.Sets.PressurePlate`
- If you are playing sounds or spawning dust in `ModTile.HitWire` or `ModTile.RightClick`, they likely don't work correctly in multiplayer. You might want to use `ModTile.HitTile` instead by calling `Wiring.HitSwitch(i, j);` and `NetMessage.SendData(MessageID.HitSwitch, number: i, number2: j);` or just `Wiring.HitSwitchAndSync(i, j);`
- `Collision.SwitchTiles` has a new `Entity` parameter and the old method is `Obsolete`, adjust MonoMod hooks or method calls as appropriate.

Tyfyter committed:
Added PressurePlate set

Tyfyter committed:
Double checked impl

direwolf420 commented:
the sample usage won't compile, PressurePlate is an int[]

JavidPack commented:
Is making a custom pressure plate useful? I guess you could make it a specific color to match a biome. Do we need an ExamplePressurePlate? Is this really the only thing needed to fully support a custom pressure plate? I had thought there might be a lot more to it from when I looked at the pressure plate code in the past. The docs will need to be updated, I still don't understand what the various values means just from reading the docs.

Tyfyter commented:
I made this PR because I noticed that two mods adding their own pressure plates for unrelated reasons had caused an incompatibility;
Adding a hook would be a good idea to make detouring the method that handles actually doing something when the pressure plate is pressed unnecessary (and could provide additional functionality in the form of letting modders disable vanilla switches via GlobalTile) but this should be all the code that's absolutely necessary for basic compatibility between mods that add pressure plates.

Tyfyter committed:
HitSwitch hook

Tyfyter commented:
Added a hook to make this useful without detouring or IL editing.

Tyfyter committed:
No longer has default implementation.

direwolf420 commented:
To clarify, it should probably be a bool return, with true indicating default behavior (which you now removed for some reason), and false for any custom behavior inside the hook

Tyfyter commented:
> To clarify, it should probably be a bool return, with true indicating default behavior (which you now removed for some reason), and false for any custom behavior inside the hook

I removed the default behavior because I felt it would be better not to have opaque default behavior controlled by magic booleans when that default behavior is just a single line the mod developer is likely already familiar with;

This way it's shorter for anything that doesn't use what was the default behavior and more readable for anything that does.

Tyfyter commented:
This indeed does not currently support weighted pressure plates.

JavidPack committed:
Merge branch '1.4.4' into pr/4774

JavidPack committed:
SwitchTiles hook, make ExampleSlopeTile a one-way sensor, docs, support proj and weighted, Example pressure plate

Chicken-Bones commented on patches/tModLoader/Terraria/Wiring.cs.patch:
```diff
@@ -74,6 +74,24 @@
+ 		if (!WorldGen.InWorld(i, j) || Main.tile[i, j] == null)
+ 			return;
+ 
++		if (!TileLoader.HitSwitch(i, j, Main.tile[i, j].TileType))
```
The docs mention something about the return value of `HitSwitch` 

```
Return true to indicate that the tile had some sort of interaction occur. This return value is only used to force specific friendly NPC to keep walking, preventing them from resting on pressure plates, for example.
```

But I don't see that reflected in the code here

JavidPack replied:
That's the `ModTile.SwitchTiles` hook docs. It's in `NPC.UpdateNPC_Inner` where `Collision.SwitchTiles` is called.

Chicken-Bones commented on ExampleMod/Content/Tiles/ExampleSlopeTile.cs:
```diff
@@ -22,5 +26,48 @@ public override bool Slope(int i, int j) {
+
+			// Check if the player is now past the bounds of this tile
+			bool success = (Main.tile[i, j].TileFrameX / 18) switch {
+				0 => playerHitbox.Left >= tileCoordinates.Right,
+				1 => playerHitbox.Top >= tileCoordinates.Bottom,
+				2 => playerHitbox.Right <= tileCoordinates.Left,
+				3 => playerHitbox.Bottom <= tileCoordinates.Top,
```
Should use `>` and `<`

JavidPack replied:
Did some testing, they need to check equals since when left and right are equal, they are no longer `Intersects`, so changing this would means that walking really slow will not trigger the tile since the player wasn't intersecting last game update.

Chicken-Bones replied:
Yes, you're correct, touching borders are not intersecting, good call.

Chicken-Bones reviewed (changes requested):
Generally looks good. The `PressurePlates` set has very specific usage, but I don't have better ideas.

direwolf420 commented:
The safest option would be to introduce a static field for this (as there is precedent like `Collision.honey`), as introducing a new parameter would break method usage in existing mods, and adding a new overload would [break MonoMod usage](https://github.com/tModLoader/tModLoader/issues/4771) as there are no overloads for `Collision.SwitchTiles` at the moment.

JavidPack committed:
Merge branch '1.4.4' into pr/4774

JavidPack committed:
Add Entity to SwitchTiles, update example, docs, hookgen

JavidPack committed:
Add Wiring.HitSwitchAndSync helper method

Chicken-Bones commented on ExampleMod/Content/Tiles/ExampleSlopeTile.cs:
```diff
@@ -62,6 +62,7 @@ public override bool SwitchTiles(int i, int j, Entity entity, Vector2 position,
 				// Since SwitchTiles is called locally, rather than calling Wiring.TripWire(i, j, 1, 1); here, we delegate that work to HitSwitch which is called on the server and all clients if we send MessageID.HitSwitch as well.
 				Wiring.HitSwitch(i, j);
 				NetMessage.SendData(MessageID.HitSwitch, number: i, number2: j);
+				// Another option is Wiring.HitSwitchAndSync(i, j), for simplicity. 
```
If we add a helper, we should use it :)

JavidPack committed:
Use helper method in EM, update docs

JavidPack committed:
Merge branch '1.4.4' into pr/4774

JavidPack committed:
Update TerrariaHooks.dll
